Lerna package code splitting using Laravel Mix

saas786 opened the pull request:
tested generated files in theme like this:
```

        function module_script( string $slug, string $script ) {

            wp_enqueue_script( $slug, get_stylesheet_directory_uri() . $script, [], Utils::get_formatted_filemtime( get_stylesheet_directory() . $script ), true );

            wp_script_add_data( $slug, 'type', 'module' );
        }

        module_script( 'cxl-ui-manifest', '/dist/js/cxl-ui/manifest.js' );
        module_script( 'cxl-ui-vendor', '/dist/js/cxl-ui/vendor.js' );
        module_script( 'cxl-ui', '/dist/js/cxl-ui/index.js' );

        // module_script( 'cxl-ui-grid', '/dist/js/cxl-ui/index-vaadin-grid.js' );
```

and in quick test it seems to work fine on local starter setup.

lkraav commented:
@saas786 could you also take a look at what can be done with https://github.com/romefrontend/rome

saas786 commented:
**Rome:**
https://github.com/romefrontend/rome
Overview:
Though I can see
https://github.com/romefrontend/rome/releases/tag/v10.0.4-beta
v10, so it should be stable enough to use, but not really.
Still seems like not even half baked.

a) Did some google research, but nothing solid found which could guide through initial hurdles.
b) Did github search, but no boilerplate or projects found using rome.
both could be my fault, not properly researched.

But after some hit and trial, these issues generated.

Official docs were misleading to get started
Ref: https://github.com/romefrontend/rome/issues/1032
Once moved further,
did this:
$ yarn rome init
result:
 pkg/package.json:2:11 projectManager/nameCollision 
  × Duplicate package name @conversionxl/cxl-ui
    1 │ {
  > 2 │   "name": "@conversionxl/cxl-ui",
      │            ^^^^^^^^^^^^^^^^^^^^
    3 │   "version": "0.2.2",
    4 │   "license": "SEE LICENSE IN LICENSE",
  i Defined already by D:\github\aybolit\packages\cxl-ui\package.json
× Found 1 problem
error Command failed with exit code 1.
`package.json` race condition results in duplicate package warnin
ref: https://github.com/romefrontend/rome/issues/955
changed
package.json name from
"name": "@conversionxl/cxl-ui"
to:
"name": "@conversionxl-t/cxl-ui"

<hr/>

Then did:
yarn rome bundle ./src/index.js ./pkg/rome/bundle.js
Result:
 $ D:\github\aybolit\node_modules\.bin\rome bundle ./src/index.js bundle.js
! This command has been hidden. Consider its usage to be experimental and do not expect support or
  backwards compatibility.
 src/index.js:1 parse/js 
  × import and export can only appear in a module
  > 1 │ import * as Headroom from 'headroom.js';
      │ ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    2 │
    3 │ export { CXLInstituteLayout } from './components/cxl-institute-layout.js';
  i Change the extension to .mjs to turn this file into a module
  i Add "type": "module" to your package.json
× Found 1 problem

<hr />

Added:
"type": "module"  to package.json

and did
yarn rome bundle ./src/index.js ./pkg/rome/bundle.js again
result:
D:\github\aybolit\node_modules\.bin\rome bundle ./src/index.js ./pkg/rome/bundle.js
! This command has been hidden. Consider its usage to be experimental and do not expect support or
  backwards compatibility.
 src/index.js:1:26 resolver/notFound 
  × Cannot find headroom.js from src/index.js
  > 1 │ import * as Headroom from 'headroom.js';
      │                           ^^^^^^^^^^^^^
    2 │
    3 │ export { CXLInstituteLayout } from './components/cxl-institute-layout.js';
 src/components/cxl-institute-layout.js:16 parse/js 
  × Unknown start to an statement expression
    14 │ const ASIDE_LOCAL_STORAGE_KEY = 'cxl-institute-layout-aside-opened';
    15 │
    16 │ @customElement('cxl-institute-layout')
 src/components/cxl-marketing-nav.js:10 parse/js 
  × Unknown start to an statement expression
     8 │ import '@vaadin/vaadin-context-menu';
     9 │
    10 │ @customElement('cxl-marketing-nav')
 src/components/cxl-vaadin-accordion.js:12 parse/js 
  × Unknown start to an statement expression
 src/components/cxl-accordion-card.js:5 parse/js 
  × Unknown start to an statement expression
    3 │ import { AccordionPanelElement } from '@vaadin/vaadin-accordion/src/vaadin-accordion-panel';
    4 │
    5 │ @customElement('cxl-accordion-card')
× Found 5 problems

<hr />

Did:
yarn rome compile ./src/index.js
result:
$ D:\github\aybolit\node_modules\.bin\rome compile ./src/index.js
! This command has been hidden. Consider its usage to be experimental and do not expect support or
  backwards compatibility.
import * as Headroom from "headroom.js";
export {CXLInstituteLayout} from "./components/cxl-institute-layout.js";
export {CXLMarketingNavElement} from "./components/cxl-marketing-nav.js";
// Order matters.
export {CXLVaadinAccordion} from "./components/cxl-vaadin-accordion.js";
export {CXLAccordionCard} from "./components/cxl-accordion-card.js";
// @todo maybe https://github.com/tc39/proposal-export-default-from
export {Headroom};
Done in 1.98s.

<hr />

**and neither "bundle" nor "compile" generated any files.**

Relevant issues:
https://github.com/romefrontend/rome/issues/128
https://github.com/romefrontend/rome/issues/955
https://github.com/romefrontend/rome/issues/1032
https://github.com/romefrontend/rome/issues/128#issuecomment-654488404

Useful article:
https://dev.to/rajikaimal/rome-a-javascript-toolchain-pad

<hr/>

My insight, I won't recommend it yet, maybe in future.

Specially when webpack v5 is shedding its old skin, and will improve on already solid system.

And laravel mix already has v6 in progress, seems like paused for now, until webpack v5 is officially released, so laravel mix will resume its development and that will be far better choice, in my opinion.

lkraav commented:
@saas786 can you update this PR to Mix v6 and see if it still works?

lkraav commented:
I think @saas786 hasn't moved on this yet. @anoblet do you think you could

* try update this PR with Mix v6 upgrade
* provide insight whether this strategy has any risks we're not aware of?

I'd also like to add another separate bundle - form elements, starting with https://github.com/vaadin/vaadin-text-field/releases/tag/v3.0.0-alpha1

lkraav commented on packages/cxl-ui/package.json:
```diff
@@ -23,7 +23,11 @@
+    "lit-html": "^1.1.2",
+    "cross-env": "~7.0.2",
+    "laravel-mix": "^5.0.4",
+    "postcss-preset-env": "^6.7.0",
```
@saas786 why do we need `postcss-preset-env` here, since we have our own SASS stack already?

lkraav commented on packages/cxl-ui/package.json:
```diff
@@ -47,6 +51,14 @@
+  },
+  "resolutions": {
+    "terser": "^4.6.11"
+  },
```
Which instruction / documentation supplied this key, what does it do?

lkraav commented on packages/cxl-ui/src/index-vaadin-grid.js:
```diff
@@ -0,0 +1,3 @@
+import * as CXLVaadinGrid from '../../cxl-vaadin-grid/src/index.js';
+
+export { CXLVaadinGrid };
```
Hm, I think we should just `import @vaadin/vaadin-grid;` here and that is all?

Why do we need `packages/cxl-vaadin-grid` separately set up?

saas786 replied:
Its more of an organization point of view, I did it like this. It can be adapted for custom js, in case we want to split it. Unless you see issue with this approach, I like to keep it.

lkraav replied:
I'll remoe it to keep things clearer in v1. Can bring it back after more people identify the need.

anoblet commented:
@leho Sure I can take a crack at it.

lkraav commented:
> Sure I can take a crack at it.

Sweet. I just posted some review comments, but not sure if @saas786 is already off for the night (Pakistan).

Maybe you can figure those q's out itmw.

This package split strategy is quite important for getting multiple important implementations unblocked. Although I think I'm just going to add `vaadin-text-field` to core bundle for now and bloat it just a bit.

saas786 commented:
> I think @saas786 hasn't moved on this yet. @anoblet do you think you could
> 
> * try update this PR with Mix v6 upgrade
> * provide insight whether this strategy has any risks we're not aware of?
> 
> I'd also like to add another separate bundle - form elements, starting with https://github.com/vaadin/vaadin-text-field/releases/tag/v3.0.0-alpha1

@lkraav Apologies I didn't get to it sooner, was planning to tackle it over the weekend.

@anoblet 
You can take a look at adding additional bundle.
> I'd also like to add another separate bundle - form elements, starting with https://github.com/vaadin/vaadin-text-field/releases/tag/v3.0.0-alpha1

 I have updated it for Mix V6, but you can do:
 
> * provide insight whether this strategy has any risks we're not aware of?

anoblet commented:
I don't believe I'm able to give recommendations at this time as I'm not familiar with laravel-mix. I was able to add cxl-vaadin-text-area to the package list though.

lkraav commented:
> I don't believe I'm able to give recommendations at this time as I'm not familiar with laravel-mix. I did add cxl-vaadin-text-area. Let me know if I did this correctly.

@anoblet it's usually prudent to amend commits with minor formatting fixes and force push your branch.

Wrt to Mix know-how, that's one vector to this effort. There's also questions like "have any alternative methods emerged recently (since we last looked), for example rollup, parcel, snowpack, ...-based".

saas786 commented:
> I did add cxl-vaadin-text-area. Let me know if I did this correctly.

@anoblet Looks fine to me.

anoblet commented:
I've worked with Webpack and Rollup before, mostly with single page applications. First thing to consider is loading the shell of the application. You do not want to split code here as it introduces unnecessary network requests which slows down the first render. Apart from that, these are the bundles that I believe we should have

* First render bundle (shell/home page)
* Component bundles
* Route bundles
* Dependency bundle (lit-element)

It seems as if laravel-mix takes care of the component/dependency bundle for us. I may have to do some research and see what would be the best approach for separating out the other bundles.

lkraav commented:
https://github.com/conversionxl/aybolit/pull/66/files is again messed up, 41 files changed.

saas786 commented:
> [#66 (files)](https://github.com/conversionxl/aybolit/pull/66/files) is again messed up, 41 files changed.

Its probably because I have rebased `syed/code-splitting` from `develop`. Ignore it for now.

lkraav commented on packages/cxl-ui/package.json:
```diff
@@ -3,7 +3,7 @@
   "author": "CXL <leho@cxl.com>",
   "license": "SEE LICENSE IN LICENSE",
-  "main": "src/index.js",
+  "main": "src/index-storybook.js",
```
What's happening here, how come cxl-ui main is something `...-storybook`?

saas786 replied:
> What's happening here, how come cxl-ui main is something `...-storybook`?

I created a separate entry point for storybook.
We have two separate bundles, one for core ui & one for playbooks.
Didn't find a way to add both here, so storybook can combine them and load both. So opted to create unique entry point for storybook.

Also here are two approaches, let me know which one you like:

Old
https://github.com/conversionxl/aybolit/commits/1a57ef8
current
https://github.com/conversionxl/aybolit/commits/e6d74e2

pawelkmpt commented:
@saas786  @lkraav it's old. can we close it?

saas786 commented:
> @saas786 @lkraav it's old. can we close it?

Yes, as related PRs already merged in relevant repos.
